aten/src/ATen/Error.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <ATen/ATenGeneral.h> // for AT_API
+
+#include <cstdint>
+#include <cstdio>
+#include <exception>
+#include <stdexcept>
+#include <string>
+
+#include <stdarg.h>
+
+namespace at {
+namespace detail {
+/// A printf wrapper that returns an std::string.
+inline std::string format(const char* format_string, ...) {
+  static constexpr size_t kMaximumStringLength = 4096;
+  char buffer[kMaximumStringLength];
+
+  va_list format_args;
+  va_start(format_args, format_string);
+  vsnprintf(buffer, sizeof(buffer), format_string, format_args);
+  va_end(format_args);
+
+  return buffer;
+}
+
+/// Represents a location in source code (for debugging).
+struct SourceLocation {
+  std::string toString() const {
+    return format("%s at %s:%d", function, file, line);
+  }
+
+  const char* function;
+  const char* file;
+  uint32_t line;
+};
+} // namespace detail
+
+/// The primary ATen error class.
+/// Provides a complete error message with source location information via
+/// `what()`, and a more concise message via `what_without_location()`. Should
+/// primarily be used with the `AT_ERROR` macro.
+struct AT_API Error : public std::exception {
+  template <typename... FormatArgs>
+  Error(
+      detail::SourceLocation source_location,
+      const char* format_string,
+      FormatArgs&&... format_args)
+      : what_without_location_(detail::format(
+            format_string,
+            std::forward<FormatArgs>(format_args)...)),
```

Comment:
Are we sure that it's ok to pass template parameter packs to C-style variadic functions like this? Looks kinda dangerous

Reply:
I think it'll compile to the correct thing (passing arguments individually)